components/Ink/src/private/MDCLegacyInkLayer.m

```diff
@@ -0,0 +1,607 @@
+    _foregroundScaleAnim.duration = kInkLayerForegroundBoundedRadiusExitDuration;
+  } else {
+    NSNumber *opacityVal = [self.presentationLayer valueForKeyPath:kInkLayerOpacity];
+    if (!opacityVal) {
+      opacityVal = [NSNumber numberWithFloat:0];
+    }
+    CGFloat adjustedDuration = kInkLayerForegroundBoundedPositionExitDuration;
+    CGFloat normOpacityVal = opacityVal.floatValue;
+    CGFloat opacityDuration = normOpacityVal / 3.f;
+    _foregroundOpacityAnim.values = @[ opacityVal, @0 ];
+    _foregroundOpacityAnim.duration = opacityDuration + adjustedDuration;
+
+    NSNumber *scaleVal = [self.presentationLayer valueForKeyPath:kInkLayerScale];
+    if (!scaleVal) {
+      scaleVal = [NSNumber numberWithFloat:0];
+    }
+    CGFloat unboundedDuration = (CGFloat)sqrt(((1.f - scaleVal.floatValue) * self.radius) /
+                                              (kInkLayerForegroundWaveTouchDownAcceleration +
+                                               kInkLayerForegroundWaveTouchUpAcceleration));
+    _foregroundPositionAnim.duration = unboundedDuration + adjustedDuration;
+    _foregroundScaleAnim.values = @[ scaleVal, @1 ];
+    _foregroundScaleAnim.duration = unboundedDuration + adjustedDuration;
+  }
+
+  _foregroundOpacityAnim.keyTimes = @[ @0, @1 ];
+  if (_foregroundOpacityAnim.duration < _foregroundScaleAnim.duration) {
+    _foregroundScaleAnim.delegate = self;
+  } else {
+    _foregroundOpacityAnim.delegate = self;
+  }
+
+  _foregroundOpacityAnim.timingFunction =
+      [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionLinear];
+  _foregroundPositionAnim.timingFunction = [self logDecelerateEasing];
+  _foregroundScaleAnim.timingFunction = [self logDecelerateEasing];
+
+  [CATransaction begin];
+  if (completionBlock) {
+    __weak MDCLegacyInkLayerForegroundRipple *weakSelf = self;
+    [CATransaction setCompletionBlock:^{
+      MDCLegacyInkLayerForegroundRipple *strongSelf = weakSelf;
+      if (strongSelf.rippleState != kInkRippleCancelled) {
+        completionBlock();
+      }
+    }];
+  }
+  [self addAnimation:_foregroundOpacityAnim forKey:kInkLayerForegroundOpacityAnim];
+  [self addAnimation:_foregroundPositionAnim forKey:kInkLayerForegroundPositionAnim];
+  [self addAnimation:_foregroundScaleAnim forKey:kInkLayerForegroundScaleAnim];
+  [CATransaction commit];
+}
+
+- (void)removeAllAnimations {
+  [super removeAllAnimations];
+  _foregroundOpacityAnim = nil;
+  _foregroundPositionAnim = nil;
+  _foregroundScaleAnim = nil;
+  self.animationCleared = YES;
+}
+
+@end
+
+static CGFloat const kInkLayerBackgroundOpacityEnterDuration = 0.6f;
+static CGFloat const kInkLayerBackgroundBaseOpacityExitDuration = 0.48f;
+static CGFloat const kInkLayerBackgroundFastEnterDuration = 0.12f;
+static NSString *const kInkLayerBackgroundOpacityAnim = @"backgroundOpacityAnim";
+
+@interface MDCLegacyInkLayerBackgroundRipple ()
+@property(nonatomic, strong) CAKeyframeAnimation *backgroundOpacityAnim;
```

Comment:
Nullability annotations.

Reply:
This is legacy code. Will adding nullability break clients?

Reply:
Done. Added nullability here.